### ISSUE TITLE
CodeReloader doesn't reload code for different instances of the same app running simultaneously

### ISSUE THREAD

**Issue by dzhlobo** [phoenixframework/phoenix#4292]
If two (or more, I guess) instances of the same phoenix app share the same `_build` folder, Phoenix CodeReload is able to reload code only for one of them, whichever gets a request first (whichever execute CodeReloader plug first). The code for another instance doesn't get updated, even though the recompiled code lies in `_build` folder ready to be loaded.

I prepared a demo repository which emulates the problem: [https://github.com/dzhlobo/phoenix-code-reloader-in-cluster](https://github.com/dzhlobo/phoenix-code-reloader-in-cluster#phoenix-codereloader-doesnt-reload-code-for-different-instances-of-the-same-app-running-simultaneously).

### Background

I have a dockerized environment for local development. I'm running two instances of the app, one serves http requests, another one processes background jobs (I use Oban). It's a single codebase, not an umbrella app. I want to have them as a single app so I can easily use PubSub, e.g. send messages from a worker running on one instance to a LiveView running on another instance.

### Investigation

**Disclaimer:** I'm not an experienced Elixir developer, I'm building my first app. Please forgive me possibly not correct explanation and feel free to correct me, I'm happy to learn.

I noticed that my workers' code doesn't get reloaded when I change the code in editor. So I went to study how Phoenix CodeReloader works. I realized that, since my workers instance doesn't have web endpoints, CodeReloader will never be invoked and there is no way to automatically trigger code reloading. But in order to debug, I connected to workers node with iex and tried to call `Phoenix.CodeReloader.Server.reload!(AppWeb.Endpoint)` manually. The code was not reloader even after this. So I studied how CodeReloader Server works.

From what I understood:

1. Whenever an app gets a request a [CodeReloader plug](https://github.com/phoenixframework/phoenix/blob/d3cd5416e5e0ee195db3708e210f804c3804d2bd/lib/phoenix/code_reloader.ex#L57) intercepts it and sends the `reload!` message to `Phoenix.CodeReloader.Server`.
2. In its turn, it [runs](https://github.com/phoenixframework/phoenix/blob/d3cd5416e5e0ee195db3708e210f804c3804d2bd/lib/phoenix/code_reloader/server.ex#L75) `mix_compile/3`.
3. [`mix_compile/3`](https://github.com/phoenixframework/phoenix/blob/d3cd5416e5e0ee195db3708e210f804c3804d2bd/lib/phoenix/code_reloader/server.ex#L170) recompiles dependencies if needed, runs some checks to throw a warning when config file was changed, and eventually comes to compiling our Phoenix app.
4. The recompilation and code reloading flow, simplified, looks like this:
  a) Get all .beam files in `_build/dev/lib/app/consolidated`.
  b) For each .beam file we run `:code.purge(module)` and `:code.delete(module)`. I'm not familiar with Erlang's code loader and code updater but from what I understood, it somehow detects what .beam files get stale, mark them, unload them, and kill processes that run this code. I don't know how it's technically possible for Erlang to detect which .beam modules became stale even before we recompile our files with elixir compiler but `:code.delete/1` indeed "delete" (perhaps not from filesystem but from memory) some .beam files while not touching others. I don't fully understand this part.
  c) When Erlang purged and deleted some .beam modules, we use `Code.delete_path/1` to remove `_build/dev/lib/app/consolidated` directory from lookup path for modules.
  d) Then we run mix compile tasks for all the compilers we use in our app. Speaking about elixir code, [mix compile.elixir invokes](https://github.com/elixir-lang/elixir/blob/06f155eb49c29e2ce97034d656f31bb8f0aa5c6e/lib/mix/lib/mix/compilers/elixir.ex#L32) elixirc compiler and it compiles new .beam modules while also generating a manifest file with some meta information about module dependencies. The manifest file has binary format so I don't know how it works yet (would be happy to learn).
  e) When .beam files are recompiled by mix, we run `Code.prepend_path/1` to add `_build/dev/lib/app/consolidated` back to the list of paths that Erland uses to find code.

So whichever app first execute `:code.purge` and `:code.delete` gets its processes restarted with a new code. E.g. if I change code, don't trigger any endpoint in my web instance but run `Phoenix.CodeReloader.Server.reload!(AppWeb.Endpoint)` on workers node, workers node code gets updated while the web instance is continue running on an old code. Even if I run `:code.purge` and `:code.delete` manually on another node, it can't detect any changes since .beam files in consolidated directory are up to date.

From this logic, it means code won't be reloaded even if I run 2 separate web instances that share the single _build directory.

### Workaround

I haven't tested it yet, but from what I understand, it should be solvable in dockerized environment by creating a separate volumes for `_build` folder for each docker compose service running the phoenix app.

I'm not sure how to solve it if you don't use docker compose.

### Ideas

I would be happy to contribute a fix for CodeReloader but I want to discuss it first.

1. Maybe, you want to discourage running 2 separate instances of the app that share single `_build` folder, we may add this to documentation. Maybe, I'm missing something and I should have used an umbrella app instead (I want my app parts to utilize Phoenix PubSub and just wanted to keep it simple). I would be happy to describe it in documentation.
2. Maybe, we can connect together `Phoenix.CodeReloader.Server` processes that are running in the same cluster and are responsible for the same `_build` folder (how to detect that folder is the same? 🤔) so only one CodeReloader.Server recompiles the code while others just do "full restart" (we will loose processes' in memory states but we do it anyway while reloading code, right?) when compilation is finished. This won't solve the problem for people who don't put different instances in the same cluster but we can explicitly note this in the documentation.
3. Maybe, Erlang has another way to reload code besides `:code.purge` and `:code.delete` I don't know about? I haven't found any in Erlang's documentation.

As a side note, I'd like to have some kind of API in Phoenix that would allow me to mark arbitrary modules as "endpoints" (maybe there is a better term) so we can trigger code reload not only via a web request, but when we call any function in the "endpoint" module. What is the best place to discuss this?

### Environment

* Elixir version (elixir -v): Elixir 1.11.2 (compiled with Erlang/OTP 23)
* Phoenix version (mix deps): 1.5.8
* Node.js version (node -v): v14.16.0
* NPM version (npm -v): 6.14.11
* Operating system: [linux alpine, 5.10.25-linuxkit](https://github.com/dzhlobo/phoenix-code-reloader-in-cluster/blob/main/Dockerfile)

### Expected behavior

When I update the code and make a request to any instance of the app I want to see results produced by changed code.

### Actual behavior

The code is updated only for the app that get the request first.

**Comment by dzhlobo** [comment]
I did a little bit more investigation. It seems `:code.prune` and `:code.delete` for modules in consolidated directory has a little with code reloading. I'm not sure why we need it and what the consolidated directory is but I think it doesn't matter.

It seems, all the work is done by `mix compile.elixir`. This task itself goes through source files that were changed, recompiles them and reloads them using the same `:code.prune` and `:code.delete` method. It does the work on one node and puts new .beam files into ebin directory. We have no chance to recompile code on other nodes running from the same directory because for `mix compile.elixir` everything seems up to date and it therefore doesn't invoke `:code.prune` and `:code.delete` and the code doesn't get reloaded.

**Comment by josevalim** [comment]
Yes, this cannot work because they will be racing each other. The first one to compile will always have outdated code and the other one does not.

**Comment by dzhlobo** [comment]
@josevalim, are you interested in changing CodeReloader behavior to cover this case? I can prepare a pull request, but maybe it is not what you want to have since it complicates the codebase.

What do you think about my proposal to "link" CodeReloader.Servers on different nodes? When one CodeReloader start recompilation, on other nodes we subscribe to file system .beam files updates and when a .beam file gets removed, updated, or added we reload this module on a node as well? I'll discuss details on forum if you approve it in general.

Also, I'm wondering, how I can cover this with tests 🤔.

**Comment by marcandre** [comment]
I got bitten by this today. I was running `mix compile --verbose` manually to check other things, and I was expecting the `phx.server` to reload the code but it never did. Couldn't it get some internal build digest / check the mtime of the manifest to reload new `.beam` (or everything)?
